src/components/tooltipMdPanel/tooltip.js

```diff
@@ -0,0 +1,346 @@
+
+        if (autohide || mouseActive || $document[0].activeElement !== parent[0]) {
+          // When a show timeout is currently in progress, then we have to cancel it,
+          // otherwise the tooltip will remain showing without focus or hover.
+          if (showTimeout) {
+            $timeout.cancel(showTimeout);
+            setVisible.queued = false;
+            showTimeout = null;
+          }
+
+          parent.off(LEAVE_EVENTS, leaveHandler);
+          parent.triggerHandler('blur');
+          setVisible(false);
+        }
+        mouseActive = false;
+      };
+
+      var mousedownHandler = function() {
+        mouseActive = true;
+      };
+
+      // To avoid 'synthetic clicks' we listen to mousedown instead of 'click'
+      parent.on('mousedown', mousedownHandler);
+      parent.on(ENTER_EVENTS, enterHandler);
+    }
+
+    function onVisibleChanged(isVisible) {
+      if (isVisible) showTooltip();
+      else hideTooltip();
+    }
+
+    function configureWatchers() {
+      if (element[0] && 'MutationObserver' in $window) {
+        var attributeObserver = new MutationObserver(function(mutations) {
+          mutations.forEach(function(mutation) {
+            if (mutation.attributeName === 'md-visible') {
+              if (!scope.visibleWatcher) scope.visibleWatcher = scope.$watch('visible', onVisibleChanged);
+            }
+            if (mutation.attributeName === 'md-direction') {
+              updatePosition();
+            }
+          });
+        });
+
+        attributeObserver.observe(element[0], { attributes: true });
+
+        // Build watcher only if mdVisible is being used
+        if (attr.hasOwnProperty('mdVisible')) {
+          scope.visibleWatcher = scope.$watch('visible', onVisibleChanged);
+        }
+      } else { // MutationObserver not supported
+        scope.visibleWatcher = scope.$watch('visible', onVisibleChanged);
+        scope.$watch('direction', updatePosition);
+      }
+
+      var onElementDestroy = function() {
+        scope.$destroy();
+      };
+
+      // Clean up if the element or parent was removed via jqLite's .remove.
+      // A couple of notes:
+      // - In these cases the scope might not have been destroyed, which is why we
+      // destroy it manually. An example of this can be having 'md-visible="false"' and
+      // adding tooltips while they're invisible. If 'md-visible' become true, at some
+      // point, you'd usually get a lot of inputs.
+      // - We use '.one', not '.on', because this only needs to fire once. If we were
+      // using '.on', it would get thrown into an infinite loop.
+      // - This kicks off the scope's '$destroy' event which finishes the cleanup.
+      element.one('$destroy', onElementDestroy);
+      parent.one('$destroy', onElementDestroy);
+      scope.$on('$destroy', function() {
+        setVisible(false);
+        element.remove();
```

Comment:
A lot of the new tooltip code relies on the previous tooltip logic. I studied the code to determine what could be removed and each line seemed like it was written to overcome an issue previously determined.